### DISCUSSION TITLE
What configuration style should we use?

### DISCUSSION THREAD

**Discussion by christiaangoossens** [christiaangoossens/hass-oidc-auth#6] [Polls]
## We are continuing with the current YAML structure and adding limited UI configuration later.

**See https://github.com/christiaangoossens/hass-oidc-auth/discussions/6#discussioncomment-12209060 for the conclusion, thank you for your input!**

---

For the configuration style of this integration, I see three main contenders, A, B and C (and a combination option D). I would like your input on which type you would like to see used in this integration, so please vote on this poll. The winner will be implemented in a later alpha release, before we go to the beta phase.

You are also welcome to leave a comment with your reasoning on this choice, or to clarify your voting decision.

Here are some benefits and downsides of each of the options:

## Option A: UI-based configuration
The workflow for this option would go like this:

1. Install the repository into HACS
2. Search for "OpenID Connect" or "SSO"
3. Download the latest release of this integration
4. Go to the Settings > Integrations section in Home Assistant and click "Add Integration" (like figure 1)
5. Find the "OpenID Connect" integration from the dropdown list (like figure 2)
6. Get presented with an options page with several input fields to configure your OIDC integration (like figure 3)

---

![integrations_page_discovered](https://github.com/user-attachments/assets/2261a0d4-c55d-48a1-ad9a-f99bda84cae8)
_Figure 1: Integrations page, source: https://www.home-assistant.io/getting-started/integration/_

![add_workday_integration](https://github.com/user-attachments/assets/401a76e5-314f-4128-90f6-f7762e13c425)
_Figure 2: Add integration dialog_

![workday_configure](https://github.com/user-attachments/assets/9f6d2648-c7d2-4809-9a57-9ba06f0b305f)
_Figure 3: Integration configuration dialog_

---

### Benefits

- All configuration happens within Home Assistant, with no file based configuration necessary
  - Configuring OIDC would be as simple as possible, similar to other projects such as Immich's configuration in the UI
  - This might allow novice/less-experienced users to the option to configure OIDC
- Integration will follow the modern standards within the Home Assistant community (see [ADR-0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md))
  - _Note: ADR-0010 does not specify a guideline for our use-case and lists our use case ("the configuration is static, is not discovered and relies on the user setting it up.") as a good use case for YAML. The decision on our integration type is: "All other integrations are configured via YAML or via the UI.", leaving the choice to the implementer._
- You may be able to configure multiple OIDC servers this way, with multiple entries showing up under the same integration (in the same way as multiple lights would, or multiple MQTT devices)
  - _Note: we will need to deal with username conflicts in that case and with a sub-optimal selection UI on the login screen._
- We can check your configuration before saving it within the UI, at least ensuring that connection can be made to your OIDC provider
- We can give additional instructions on how to use it within the UI, such as which URL to start at.

### Downsides

- Configuration is done within the Home Assistant database (actually a json file) and thus needs to be restored with a Home Assistant backup, while it can be restored (see https://github.com/christiaangoossens/hass-oidc-auth/discussions/6#discussioncomment-11688030, it might be harder to do so)
- The configuration screen might become very long or need multiple steps to support all use cases (or will need to be simplified losing options), such as:
  - Picking different OIDC integration methods to support different providers (Authentik, Authelia) with different methods (public/confidential) and different claims (preferred_username, username, email, nickname)
  - If you want to automatically create a person/user or not
  - Deny/allow lists for usernames/emails
  - Group configuration to automatically assign admin/user roles, or more specific roles when RBAC is implemented
  - How to deal with username conflicts/merging accounts
- Configuration might become too easy for novice users, thus leading to the issues that the HA team brought up when OIDC was originally proposed for Home Assistant
  - Users might enable OIDC, disable username/password login (especially if we make a config option for it in the UI) and then lock themselves out if they made a mistake in the configuration
  - Users might not understand the consequence of enabling OIDC for their MFA settings or configure it with a poorly configured provider/server thus allowing for unauthorized access to their smart home
  - See https://community.home-assistant.io/t/open-letter-for-improving-home-assistants-authentication-system-oidc-sso/494223#consequences-15 for more information (especially balloob's comment "With OpenID it is getting very easy to link external authentication")

## Option B: YAML-based configuration with our own 'domain' in the config
_This is currently what the pre-alpha does._

The workflow for this option would go like this:

1. Install the repository into HACS
2. Search for "OpenID Connect" or "SSO"
3. Download the latest release of this integration
4. Go to your configuration.yaml file and add the following code anywhere into your config:

```yaml
auth_oidc:
    client_id: "exampleId"
    client_secret: !secret client_secret
    discovery_url: "https://example.com"
    claims:
      username: "preferred_username"
      group: "groups"
    roles:
     admin: "admin"
     user: null
```

5. Restart Home Assistant to enable the OIDC auth provider

This configuration would add an OIDC authentication option to your providers, without removing any of the current auth options as configured in 'homeassistant.auth_providers'. It then configures a client ID and secret using the `!secret` helper such that your config can be shared. It then configures some advanced (optional) settings, such as which claims to use and which group names to map to roles in Home Assistant. 

> [!NOTE]  
> The YAML schema given above is an example and not the actual keys that will be used.

### Benefits

- Configuration is backed up in all backups (even file level backups/whole VM backups) and can be edited without opening Home Assistant, even if Home Assistant is broken/you cannot login
- There are no conflicts with the other keys in the configuration, even if Home Assistant changes the auth provider system: https://www.home-assistant.io/docs/authentication/providers/
- There is plenty of space to allow for any advanced configuration that the community wants
- If you want to disable the default authentication method, you must do so explicitly by specifying `homeassistant.providers` with an empty array.

### Downsides

- The integration must first be downloaded from the UI and then configured from YAML. There is no way to configure it from the UI directly. This might be less approachable for novice users.
  - This might make it more accessible for the Home Assistant team to adopt this feature as a native integration, as it's clearly focused on advanced users only.
- You cannot configure multiple OIDC providers easily.

## Option C: YAML-based configuration under 'homeassistant.auth_providers'
1. Install the repository into HACS
2. Search for "OpenID Connect" or "SSO"
3. Download the latest release of this integration
4. Go to your configuration.yaml file and add the following code anywhere into your config:

```yaml
homeassistant:
  auth_providers:
    - type: homeassistant
    - type: oidc
      name: "Google"
      client_id: "1234"
      client_secret: !secret google_secret
      configuration: https://accounts.google.com/.well-known/openid-configuration
      emails:
        - "john.doh@gmail.com"
    - type: oidc
      name: "Auth0"
      client_id: "1234"
      client_secret: !secret auth0_secret
      configuration: https://xxx.auth0.com/.well-known/openid-configuration
      emails:
        - "john.doh@gmail.com"
       
```

5. Restart Home Assistant to enable the OIDC auth provider

This configuration would add an OIDC authentication option to your auth providers, forcing you to specify all of them and overriding Home Assistants default config. Note that `- type: homeassistant` is required here to enable the original username/password login, leaving it out would leave OIDC as the only option to login.

> [!NOTE]  
> The YAML schema given above is an example (from https://github.com/home-assistant/core/pull/32926) and not the actual keys that will be used.

### Benefits

- This is likely the way it was originally intended to be configured when the HA team added the concept of auth providers. The native provider (username/password), trusted networks and command line also follow this pattern.
- Configuration is backed up in all backups (even file level backups/whole VM backups) and can be edited without opening Home Assistant, even if Home Assistant is broken/you cannot login.
- There is plenty of space for extra YAML keys (such as `emails` in the example above)

### Downsides

- It's very easy to lock yourself out if you forget to include the base username/password provider, as specifying `homeassistant.auth_providers` overrides the default configuration of Home Assistant (example in https://github.com/christiaangoossens/hass-oidc-auth/discussions/6#discussioncomment-11697164)
- If the Home Assistant team changes the configuration of Auth Providers, our integration will likely break as it no longer follows the same configuration system.
- It might violate the concept of [ADR-0007](https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md) which requests integrations to use their own "domain key"
```yaml
## Example from ADR-0007
# Allowed
awesome_integration:
  username: user

# Not allowed
sensor:
  - platform: awesome_integration
    username: user
```

## Option D: Both, offer novice UI setup and advanced YAML setup
Finally, we could also implement both. In this scenario, we would likely offer a very simplified UI setup (such as pre-configuring specific types like Authentik, Authelia and Keycloak and only offering those options in the UI). Novice users can then use the UI and configure their provider with only necessary fields and sane/secure defaults.

Advanced users would not use the UI and use the YAML setups specified above (depending on which gets more votes/community support).

### Benefits
- Novice users can use the UI, while YAML remains available.
- Users can make a decision to either use the YAML config with backups (infrastructure as code) or use the UI setup (more HA-like)

### Downsides
- Has a mix of downsides from the options above, such as possibly making it too easy to misconfigure
- May lead to many issues from confused users as the setup process explanation will get complicated, with too many options
- May lead to issues stating that their preferred OIDC provider is not available in the UI, while it can be configured using YAML
- Additional development time to maintain both options (I will need more contributors to this project!)

**Comment by arturpragacz** [comment]
>Configuration is done within the Home Assistant database and thus needs to be restored with a Home Assistant backup; it cannot be restored from outside Home Assistant

This is not strictly true. It is not stored in the database, but in a simple json file.

**Reply by christiaangoossens** [reply]
That's good to know, is it possible to fix that outside HA as well in an easy way on all platforms (OS, container, core) if something is broken and you cannot login without long instructions or permission errors?

**Reply by arturpragacz** [reply]
Yes, it can always be edited. I would however strongly advise against recommending it to users. That file is not meant to be edited by hand, users could easily seriously mess their Home Assistant by doing so.

If you want to have some sort of recovery mode, just simply add it to YAML configuration instead. Something like this should suffice:
```yaml
auth_oidc:
  disable: true
```

**Reply by christiaangoossens** [reply]
@arturpragacz Hmm, that worries me. Recovery mode (disabling it) isn't good enough if people didn't make a backup or have backup credentials.

If we go the UI route, it needs to be editable if you cannot login to HA at all because you don't have a backup option, without major risks. 
I mostly want to tackle scenario's like their client secret expiring or HA being inaccessible due to other changes at the OIDC provider.

While I don't recommend you disable the username/password login, we need to be able to solve issues for people who did do that or simply forgot the username/password because they never use it. People who need this (and didn't make a backup or have backup credentials in the first place), most likely won't be the ones who safely edit the JSON file, I imagine.

Seems like a major downside for any UI based setup method for this integration to me.

**Reply by arturpragacz** [reply]
You keep thinking about it from a non-HA-native perspective and I think this is a mistake.

>if people didn't make a backup or have backup credentials.

They can always just [reset their password](https://www.home-assistant.io/docs/locked_out/#forgot-password).

>While I don't recommend you disable the username/password login, we need to be able to solve issues for people who did do that

I'm not sure how you imagine people are going to disable username/password login, but if it's your integration doing it, then the simple YAML solution I gave above solves that problem.

>People who need this (and didn't make a backup or have backup credentials in the first place), most likely won't be the ones who safely edit the JSON file, I imagine. Seems like a major downside for any UI based setup method for this integration to me.

People who can't edit the YAML file in order to recover from misconfiguration would not use YAML-only integration in the first place. You very strongly limit your reach that way. One advantage of the UI approach is that if you can detect that the configuration given by the user doesn't work, you can prevent him from saving it in the first place.

**Reply by christiaangoossens** [reply]
@arturpragacz Thanks for your quick response! It's good to see a more HA native way of thinking.

> You keep thinking about it from a non-HA-native perspective and I think this is a mistake.

I'm definitely not an HA native, more of a homelabber/self-hoster, so HA is just one of the tens of things I run. Most of the self-hosted tools do this using configuration by code (either yaml/env vars etc). I'm here from the OIDC experience side, having worked with the spec and in security for years, very much not from the HA experience or community side, so that's what this (and future) discussion post is for.

Remember to vote on the poll, I will just pick the most popular one in a few weeks (or implement both), there are no votes for UI yet. I expect the audience for this integration to be split between HA users that dabble in SSO and more hardcore homelabbers who don't mind YAML, so I would like to hear everyone's opinions on it before deciding on an implementation.

> They can always just [reset their password](https://www.home-assistant.io/docs/locked_out/#forgot-password).

True, doesn't look fun but it's possible.

> I'm not sure how you imagine people are going to disable username/password login

I don't think we should offer that option at all, both in YAML and in UI, but if you wanted it, you can do it using the following YAML:

```yaml
homeassistant:
  auth_providers:
```

Effectively, set `homeassistant.auth_providers` to be an empty list, thus removing the default `homeassistant` auth provider. Arguably, people who did this themselves can also revert that in YAML. No involvement from this integration here at all.

> You very strongly limit your reach that way.

Yeah, that's one of the arguments for and against it, judging from the last two years of responses on the forum. OIDC/SSO is not only an entreprise tool, but it's a bit of a poweruser tool non the less. While I encourage everyone to try and learn it, it should not be regarded on the same level as many of the other HA click and use integrations and it requires a bit more reading and figuring things out.

Although, I am including a guide on https://github.com/stonith404/pocket-id (simple passkey OIDC for more novice users) as well, besides Authentik and Authelia, so the audience for homelab might become a bit more novice-friendly in the future.

> One advantage of the UI approach is that if you can detect that the configuration given by the user doesn't work, you can prevent him from saving it in the first place.

I will add this to the post as an edit, it's a great point in favor of UI based.

**Comment by patienttruth** [comment]
While I don't have the skill to contribute to the code I'd vote for UI setup, however if it can't be repaired without Home Assistant running I'd say YAML. If the json file Artur mentioned can be modified than it's a moot point.

All that said, restoring HA from a backup isn't a huge deal, so perhaps GUI all the way is best? I'm sure it will get much wider adoption the easier it is to implement.

Much gratitude to you for stepping up to fill this need.

**Reply by christiaangoossens** [reply]
Thanks for your input! Make sure to also vote in the poll itself.

Wider adoption is a bit of a hard point if you read the community discussion over the last two years. 

On the one hand, I agree it should be as easy as possible, to make sure that everyone who wants it can use it easily. On the other hand, there was a lot of pushback as well to the idea of SSO/OIDC and many people also viewed it as a silver bullet to get a lot of other things done, such as RBAC, Enterprise login, WebAuthN. Some even reacted to feel forced to use this to just get a passkey or better MFA.

While all of that can be configured through your OIDC provider, it seems to me that those require some detailed setup anyway, often through config files of their own. Not really the path I would like users that don't feel comfortable editing yaml should go down, for their own safety/preventing config mistakes. SSO is not an easy hobby.

How do you view that point?

**Comment by Node815** [comment]
My observation of the direction the Home Assistant team is moving over the last year is they are pushing more to the UI for configuration and replacing some of the older configuration.yaml entries.    I think having the option to use the UI will make it more consistent and easier to implement.

**Comment by jsimcina** [comment]
@christiaangoossens most of the cons and the care here is the prevention of users locking themselves out.  Why not take an approach similar to other SAML/OAuth solutions and add a querystring parameter to bypass SSO and fall back to HAs native login process?  Like a ?SkipSSO=true thrown on to the URL?

**Reply by christiaangoossens** [reply]
I'm afraid you misunderstood the issue. When using the integration using the default settings, normal HA login will always be available on the login screen and secure defaults are used. Therefore, there would be no need for a specific URL to fallback to local login.

The issue here lies in users blindly copying configuration or searching for methods to accomplish things they don't yet fully understand. Users themselves, through various methods, can disable normal HA login, misconfigure user linking and/or create an implicit dependance on their OIDC user login (especially if RBAC is introduced someday or if they forget their normal HA credentials). With security integrations, just 'having it work' does not guarantee your service is also secure or recoverable if something bad happens.

You can see this principle happening in the discussions already, where many examples have `username_linking` enabled, even though that is a dangerous setting to use if you don't understand the impact fully. Therefore, it is disabled by default if you don't specify it, but users might not know that.

Previous comments in this discussion post have however changed my stance, I was of the opinion that YAML only might force novice users to discover more information first, but the opposite might actually be true (with copy and pasting). Additionally, the UI allows us to always show warnings in place and to verify the configuration before saving. I'm leaning towards adopting option D for that reason, with some niche settings in YAML, but the general configuration route in the UI.

**Comment by nyawox** [comment]
As a believer in infrastructure as a code, please, please keep the yaml configuration as an option, even if it end up undocumented. Sorry to be that guy but I want to just deploy with single command and be done with minimal manual interactions and this is becoming more and more difficult as the core team move to UI configurations. imo what really should be done (and I enjoy) in the UI is lovelace dashboard configurations, rather than integrations which often require provisioning of secrets. but the team seem to be moving backwards... before writing this i was attempting to rewrite an entire custom component to use yaml instead of config flow and as someone who don't main python it was too frastruating that i've just gave up and opted to go back calling rest api with my custom script. I believe most people who can figure out what pkce means are also capable of writing yaml. but ofc if you have the capacity, making GUI the recommended option might be the right direction...?

**Comment by jsimonetti** [comment]
I can't really comment much on the HA native way, other then it creates a chicken-egg problem (and a securit issue, see below). To be able to configure the oidc provider in the UI, an administrator account in the local database is required.
In general system administration, it is best practice to not have a local administrator user.
The only way to avoid this problem is by using yaml configuration. 

Another downside of using ha ui style configuration (and the main reason I am a very outspoken hater of the move towards ui) is the storage of secret data in a plain text file. With all things configured through the ui, secret information such as usernames, passwords, (or in oidc case) client-id and client-secret are stored in plain text in a file on the filesystem.
This completely disabled the possibility of securing secret information and goes against every basic rule of security I know of.

**Reply by christiaangoossens** [reply]
Thank you for your input. Easily being able to fix your configuration to restore access is a big plus for the YAML config. 

Also, note that you don't need a client secret, as PKCE with the redirect URI (public client) already ensures we return the token to the original app, provided that your OIDC provider enforces the use of PKCE correctly. For many home purposes, this is easily secure enough. That prevents the need to store it in plaintext. Client ID is never a secret in the first place.

**Reply by Th0masDB** [reply]
But why wouldn't we just add an extra layer of protection if it is not really difficult to add? I think it is worth it to use PKCE and a client secret if it can be stored not in plain text file.

**Reply by christiaangoossens** [reply]
Well, because HA doesn't support not having it in plain text afaik (either in the json or in secrets yaml). Although I think it's a bit overkill to worry about that (physical access to your machines).

Also, having a client secret isn't always better if you leak it, having a leaked client secret is worse than just public client as you will get more trust from the OIDC provider.

In the end, the choice on if you want to use a client secret and if you are sure that you can keep it safe, is up to the user.

**Reply by jsimonetti** [reply]
When you use a secrets file (in yaml mode) this file can be stored encrypted and decrypted only for the running process (for example using systemd credentials).
This is completely out of scope here and a general sysadmin challenge where multiple tools exist that can cover this part.

**Reply by christiaangoossens** [reply]
@jsimonetti True, good point. I think it's clear from the discussion that yaml should stay an option at least, so don't worry.

**Comment by christiaangoossens** [comment]
Okay, conclusions time. It seems from the voting and discussion that the option of keeping the current YAML as the main configuration, but adding (limited) UI options later on to enable novice users to configure more generic setups (Authentik, Authelia, Pocket-ID) would be a good route.

I am going to close this discussion with that note and add an open TODO item to add that limited UI configuration.
